### PR TITLE
JW-168 Implement malloc, free, realloc, zalloc for windows kernel

### PR DESCRIPTION
**Pull request by mmasztalerczuk**

### CODE REVIEW

File: library/c/sandesh.h [hunk 1]
```diff
@@ -25,6 +25,52 @@ extern "C" {
...
+    return mem;
+}
+
+static void
+s_free(void *mem)
+{
+    if (mem) {
+        mem = (char*)mem - sizeof(unsigned int);
+        ExFreePoolWithTag(mem, 'RVCO');
+    }
```
Comment (m-kostrzewa) [inline library/c/sandesh.h]: I think having ```else DbgPrint``` here would be useful for catching errors

File: library/c/sandesh.h [hunk 2]
```diff
@@ -25,6 +25,52 @@ extern "C" {
 #include <Wsk.h>
 #include <Ws2ipdef.h>
 
+static void *
+s_malloc(unsigned int size)
+{
+    void *mem = ExAllocatePoolWithTag(NonPagedPool, size+1, 'RVCO');
+    *(int*)mem = size;
+    mem = (char*)mem + sizeof(unsigned int);
```
Comment (m-kostrzewa) [inline library/c/sandesh.h]: Why add (and subtract in other functions) sizeof uint?

**Review by m-kostrzewa — APPROVED** [review]
couple questions

File: library/c/sandesh.h [hunk 3]
```diff
@@ -25,6 +25,52 @@ extern "C" {
 #include <Wsk.h>
 #include <Ws2ipdef.h>
 
+static void *
+s_malloc(unsigned int size)
+{
+    void *mem = ExAllocatePoolWithTag(NonPagedPool, size+1, 'RVCO');
```
Comment (sodar) [inline library/c/sandesh.h]: `size+1` will not allocate enough space for header and requested buffer size. `size` will be 4 bytes, so you need at least 4 bytes for header.

File: library/c/sandesh.h [hunk 4]
```diff
@@ -25,6 +25,52 @@ extern "C" {
...
+    *(int*)mem = size;
+    mem = (char*)mem + sizeof(unsigned int);
+    return mem;
+}
+
+static void *
+s_zalloc(unsigned int size)
+{
+    void *mem = s_malloc(size);
+    NdisZeroMemory(mem, size);
```
Comment (sodar) [inline library/c/sandesh.h]: According to MSDN `RtlZeroMemory` is preferred (see **Requirements** in https://msdn.microsoft.com/en-us/library/windows/hardware/ff564698(v=vs.85).aspx)

File: library/c/sandesh.h (hunk 3 above)
Comment (sodar) [inline library/c/sandesh.h]: You can move `'RVCO'` tag into a global variable as in `windows/vrouter_mod.c` in vRouter code.

```
ULONG  SxExtAllocationTag = 'RVCO';
```

File: library/c/sandesh.h [hunk 5]
```diff
@@ -25,6 +25,52 @@ extern "C" {
...
+
+    return mem;
+}
+
+static void
+s_free(void *mem)
+{
+    if (mem) {
+        mem = (char*)mem - sizeof(unsigned int);
+        ExFreePoolWithTag(mem, 'RVCO');
```
Comment (sodar) [inline library/c/sandesh.h]: Same as above in `s_malloc`.

**Review by sodar — CHANGES_REQUESTED** [review]

File: library/c/sandesh.h (hunk 3 above)
Comment (mmasztalerczuk) [reply library/c/sandesh.h]: 👍

File: library/c/sandesh.h (hunk 4 above)
Comment (mmasztalerczuk) [reply library/c/sandesh.h]: 👍

File: library/c/sandesh.h (hunk 5 above)
Comment (mmasztalerczuk) [reply library/c/sandesh.h]: :+1:

File: library/c/sandesh.h [hunk 6]
```diff
@@ -25,6 +24,54 @@ extern "C" {
...
+static void *
+s_realloc(void *mem, unsigned int size)
+{
+    unsigned int old_size;
+
+    void *mem_tmp = s_malloc(size);
+    mem = (char*)mem - sizeof(unsigned int);
+    old_size = *(unsigned int*)mem;
+    RtlCopyMemory(mem_tmp, mem, old_size);
+    s_free(mem);
```
Comment (sodar) [inline library/c/sandesh.h]: You have modified `mem` variable before calling `s_free`. It might cause a bug, because `s_free` also modifies it.

**Review by sodar — CHANGES_REQUESTED** [review]

File: library/c/protocol/thrift_binary_protocol.c [hunk 7]
```diff
@@ -18,7 +18,10 @@
  */
 
 #include "sandesh.h"
-
+#pragma warning(disable: 4100)
+#pragma warning(disable: 4018)
+#pragma warning(disable: 4701)
```
Comment (mclapinski) [inline library/c/protocol/thrift_binary_protocol.c]: maybe include "vr_os.h"?

File: library/c/protocol/thrift_binary_protocol.c [hunk 8]
```diff
@@ -18,7 +18,10 @@
...
+unsigned short _fltused = 0;
```
Comment (mclapinski) [inline library/c/protocol/thrift_binary_protocol.c]: put this in ifdef _WINDOWS

File: library/c/sandesh.h (hunk 2 above)
Comment (m-kostrzewa) [reply library/c/sandesh.h]: @mmasztalerczuk explain please

File: library/c/sandesh.h [hunk 9]
```diff
@@ -25,6 +24,53 @@ extern "C" {
...
+    return mem_tmp;
+}
+
+#define os_malloc(size)                  s_malloc(size)
+#define os_zalloc(size)                  s_zalloc(size)
+#define os_free(ptr)                     s_free(ptr)
+#define os_realloc(ptr, size)            s_realloc(ptr, size)
+#define os_log(level, format, ...)
```
Comment (mclapinski) [inline library/c/sandesh.h:72]: So this should happen if not defined _NTKERNEL

File: library/c/sandesh.h [hunk 10]
```diff
@@ -94,11 +140,6 @@ extern int vrouter_dbg;
 #define OS_LOG_ERR LOG_ERR
 #define OS_LOG_DEBUG LOG_DEBUG
 
-#define os_malloc(size)                  malloc(size)
-#define os_zalloc(size)                  calloc(1, size)
-#define os_realloc(ptr, size)            realloc(ptr, size)
-#define os_free(ptr)                     free(ptr)
-#define os_log(level, format, ...)       winsyslog(level format, __VA_ARGS__)  //only for now
```
Comment (mclapinski) [inline library/c/sandesh.h:101]: this should happen if defined _NTKERNEL

**Review by mclapinski — CHANGES_REQUESTED** [review]

**Review by sodar — APPROVED** [review]

**Pull request merged**